snuba_sdk/query_validation.py

```diff
@@ -0,0 +1,224 @@
+
+    Ensure that all the columns referenced in the query are in the data model for that entity.
+
+    :param match: The Entity of the query.
+    :type match: Entity
+    :param all_columns: All the columns referenced in the query.
+    :type all_columns: Set[Expression]
+    """
+    for column in all_columns:
+        assert isinstance(column, Column)
+        column.validate_data_model(match)
+
+
+class RequiredColumnError(Exception):
+    pass
+
+
+def _check_entity_has_required_columns(
+    entity: Entity, conditions: Optional[Sequence[Union[BooleanCondition, Condition]]]
+) -> None:
+    """
+    For a given entity, flatten the AND conditions to find just the top level ones and
+    make sure there is a correct condition for each required column and the required
+    time column. Conditions must be on a particular op(s). This doesn't handle ORs
+    completely, since it's possible to combine ORs in a way to have valid conditions.
+
+    :param entity: The entity to check for required columns
+    :type entity: Entity
+    :param conditions: The conditions of the query
+    :type conditions: Optional[Sequence[Union[BooleanCondition, Condition]]]
+    :raises RequiredColumnError
+    """
+    if not entity.data_model:
+        return
+
+    if not conditions:
+        raise RequiredColumnError("where clause is missing required columns")
+
+    schema = entity.data_model
+    top_level = get_top_level_conditions(conditions)
+
+    required_matches: List[Tuple[Set[Op], ColumnModel]] = [
+        *[({Op.EQ, Op.IN}, req_column) for req_column in schema.required_columns],
+    ]
+    not_matched = set(range(len(required_matches)))
+    for i, match in enumerate(required_matches):
+        if i not in not_matched:
+            continue
+
+        ops, col_to_match = match
+        if _find_matching_condition(ops, col_to_match, top_level, entity.alias):
+            not_matched.remove(i)
+
+    if not_matched:
+        entity_alias = f"{entity.alias}." if entity.alias else ""
+        missing = sorted(
+            [f"'{entity_alias}{required_matches[i][1].name}'" for i in not_matched]
+        )
+        s = "s" if len(missing) > 1 else ""
+        raise RequiredColumnError(
+            f"where clause is missing required condition{s} on column{s} {', '.join(missing)}"
+        )
+
+    ops = {Op.GTE, Op.LT}
+    missing_ops = set()
+    for op in ops:
+        if not _find_matching_condition(
+            {op}, schema.required_time_column, top_level, entity.alias
+        ):
+            missing_ops.add(op)
+
+    if missing_ops:
+        s = "s" if len(missing_ops) > 1 else ""
+        entity_alias = f"{entity.alias}." if entity.alias else ""
+        exc_value = [f"{op.value}" for op in missing_ops]
+        raise RequiredColumnError(
+            f"where clause is missing required {', '.join(exc_value)} condition{s} on column '{entity_alias}{schema.required_time_column.name}'"
+        )
+
+
+def _find_matching_condition(
```

Comment:
This function's name is also misleading. `_find_matching_condition` implies that there will be a condition returned.